opengovsg/postmangovsg
# feat: Encrypted database backup lambda function

## Problem

Implements a lambda function that create a backup of a Postgresql database by running `pg_dump` and encrypting its output with a symmetric encryption algorithm. 

Closes #703 

## Solution

**Features**:
- Created a lambda layer that installs and package up `pg_dump` and it's dependencies so that it can be used within the lambda runtime
- Implement lambda to run, encrypt and upload `pg_dump` output to a S3 bucket

## Tests
- Run `database-backup-staging` lambda through the console using `RDSEvent`
- Download the backup files from `postmangovsg-rds-backup-staging` S3 bucket. The folder name follows the following format `YYYY-MM-DD`.
- Build and decrypt the files using `scripts/decrypt.ts`. You can choose to either use the staging backup private key in our 1Password or GCP KMS. (See below for detailed instructions) 
- Test out the dump by running `pg_restore` and inspecting the content of the decrypted `secrets.dump`

**Using local private key for decryption**
Download staging private key PEM from 1Password (Staging - Backup Encryption Key). Download the dump files from S3 and run the following to decrypt the pgdump (`postman.dump`). 

```bash
$ node build/scripts/decrypt.js \
    --privateKey=key.pem \
    --input postman.dump \
    --output postman.decrypted.dump \
    --params postman.json
```
Repeat again for `secrets.dump`. 

**Using GCP KMS for decryption**
Ensure you have [installed and setup](https://cloud.google.com/sdk/docs/install) the gcloud CLI (`gcloud init`). 

Download the GCP service account credentials from 1Password (Staging - GCP KMS Decrypter Credentials). Download the dump files from S3 and run the following to decrypt the pgdump (`postman.dump`). 

```bash
$ export GOOGLE_APPLICATION_CREDENTIALS=[path to the JSON key file]
$ export KEY_ID=[copy key ID from 1Password]
$ node build/scripts/decrypt.js \
    --kmsKey=$KEY_ID \
    --input postman.dump \
    --output postman.decrypted.dump \
    --params postman.json
```
Repeat again for `secrets.dump`. 

## Deploy Notes

**New environment variables**:
- `PG_VERSION` : The current version of Postgresql that we're running. Used to determine the version of `pg_dump` that we will be packaging with the lambda.
- Staging
   - `STAGING_BACKUP_ROLE` : Backup lambda IAM role ARN
   - `STAGING_BACKUP_DB_URI` : Database URI to use for `pg_dump`. Use read replica.
   - `STAGING_BACKUP_BUCKET_NAME` : S3 bucket name to store backup in
   - `STAGING_BACKUP_PUBLIC_KEY` : Public key used to encrypt encryption key
- Production
   - `PRODUCTION_BACKUP_ROLE` : Backup lambda IAM role ARN
   - `PRODUCTION_BACKUP_DB_URI` : Database URI to use for `pg_dump`. Use read replica.
   - `PRODUCTION_BACKUP_BUCKET_NAME` : S3 bucket name to store backup in
   - `PRODUCTION_BACKUP_PUBLIC_KEY` : Public key used to encrypt encryption key

**Infrastructure updates**
- [ ] Standardise backup window for all database to between 3:00AM - 3:30AM SGT
- [ ] Set CloudWatch alarm for backup lambda failure
- [x] Generate and [import new key in GCP KMS](https://cloud.google.com/kms/docs/importing-a-key) for production backup

## TODOs
- [x] Create IAM role for backup lambda with necessary permissions (RDS, S3)
- [x] Update Travis to build and package lambda with `pg_dump`
- [x] ~~Implement the use of custom KMS CMK for generation of data key for symmetric encryption~~
   - _Ciphertext generated by KMS is [not portable](https://docs.aws.amazon.com/kms/latest/developerguide/importing-keys.html#importing-keys-considerations). Even if you have the same key material, decryption is not possible outside of KMS_ 
- [x] Asymmetric encrypt encryption key for each dump
- [x] Add dumping for SecretsManager
- [x] Setup SNS queue and connect up RDS events to trigger lambda
